src/qibolab/result.py

```diff
@@ -1,10 +1,165 @@
+    """
+
+    def __init__(self, data: np.ndarray):
+        self.voltage: npt.NDArray[np.complex128] = data
+
+    def __add__(self, data):
+        return self.__class__(np.append(self.voltage, data.voltage))
+
+    @property
+    def voltage_i(self):
+        """Signal magnitude in volts."""
+        return self.voltage.real
+
+    @property
+    def voltage_q(self):
+        """Signal magnitude in volts."""
+        return self.voltage.imag
+
+    @cached_property
+    def magnitude(self):
+        """Signal magnitude in volts."""
+        return np.sqrt(self.voltage_i**2 + self.voltage_q**2)
+
+    @cached_property
+    def phase(self):
+        """Signal phase in radians."""
+        return np.angle(self.voltage_i + 1.0j * self.voltage_q)
+
+    @property
+    def serialize(self):
+        """Serialize as a dictionary."""
+        serialized_dict = {
+            "MSR[V]": self.magnitude.flatten(),
+            "i[V]": self.voltage_i.flatten(),
+            "q[V]": self.voltage_q.flatten(),
+            "phase[rad]": self.phase.flatten(),
+        }
+        return serialized_dict
+
+    @property
+    def average(self):
+        """Perform average over i and q"""
+        average_data = np.mean(self.voltage, axis=0)
+        std_data = np.std(self.voltage, axis=0, ddof=1) / np.sqrt(self.voltage.shape[0])
+        return AveragedIntegratedResults(average_data, std_data)
+
+
+class AveragedIntegratedResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.INTEGRATION and AveragingMode.CYCLIC
+    or the averages of ``IntegratedResults``
+    """
+
+    def __init__(self, data: np.ndarray, std: np.ndarray = np.array([])):
+        super().__init__(data)
+        self.std: Optional[npt.NDArray[np.float64]] = std
+
+    def __add__(self, data):
+        new_res = super().__add__(data)
+        new_res.std = np.append(self.std, data.std)
+        return new_res
+
+
+class RawWaveformResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.SINGLESHOT
+    may also be used to store the integration weights ?
+    """
+
+
+class AveragedRawWaveformResults(AveragedIntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.CYCLIC
+    or the averages of ``RawWaveformResults``
+    """
+
+
+# FIXME: If probabilities are out of range the error is displeyed weirdly
```

Comment:
By this you mean if the probabilities are not between 0 and 1?

Reply:
Yes, I was checking if they were there before but not anymore and I forgot to remove that. So now that we are on this, should I be cheking porbabilities to be on that range ?

Reply:
This probabilities are computed using shots, right?
If so they should always be bounded between 0 and 1 I think.

Reply:
They should if the devices are working as they should

Reply:
I will not check you will probably see it on the data